### PR TITLE
Excessive memory consumption in slicing code

### LINKED ISSUE #938: Excessive memory consumption in slicing code
## Current Behavior
Allocating MaxObjectSize per incoming request.

## Expected Behavior
Doing something more sane when dealing with smaller requests.

## Context
#931.

### PR DESCRIPTION
**Pull request by smallhive**
Closes #938

### CODE REVIEW

File: api/layer/multipart_upload.go [hunk 1]
```diff
@@ -312,7 +312,19 @@ func (n *layer) uploadPart(ctx context.Context, multipartInfo *data.MultipartInf
 		},
 	}
 
-	chunk := n.buffers.Get().(*[]byte)
+	var (
+		chunk          *[]byte
+		isReturnToPool bool
+	)
+
+	if p.Size > n.neoFS.MaxObjectSize() {
```
Comment (roman-khimov) [inline api/layer/multipart_upload.go:320]: Maybe `n.neoFS.MaxObjectSize()/2`? 63M request allocating below would be worse than the same request reusing 64M buffer.

File: internal/neofs/neofs.go [hunk 2]
```diff
@@ -335,8 +335,20 @@ func (x *NeoFS) CreateObject(ctx context.Context, prm layer.PrmObjectCreate) (oi
...
 		opts.SetPayloadSize(prm.PayloadSize)
 
-		data := x.buffers.Get()
-		chunk := data.(*[]byte)
+		var (
+			chunk        *[]byte
+			returnToPool bool
+		)
+
+		if prm.PayloadSize > 0 && prm.PayloadSize < uint64(x.MaxObjectSize()) {
```
Comment (roman-khimov) [inline internal/neofs/neofs.go:343]: Same here.

File: internal/neofs/neofs.go [hunk 3]
```diff
@@ -373,11 +387,24 @@ func (x *NeoFS) CreateObject(ctx context.Context, prm layer.PrmObjectCreate) (oi
...
 	}
 
-	data := x.buffers.Get()
-	chunk := data.(*[]byte)
+	var (
+		chunk        *[]byte
+		returnToPool bool
+	)
+
+	if prm.PayloadSize > 0 && prm.PayloadSize < uint64(x.MaxObjectSize()) {
```
Comment (roman-khimov) [inline internal/neofs/neofs.go:395]: And here.

File: api/layer/multipart_upload.go (hunk 1 above)
Comment (smallhive) [reply api/layer/multipart_upload.go]: Updated

File: internal/neofs/neofs.go (hunk 2 above)
Comment (smallhive) [reply internal/neofs/neofs.go]: Updated

File: internal/neofs/neofs.go (hunk 3 above)
Comment (smallhive) [reply internal/neofs/neofs.go]: Updated

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**